### PR TITLE
#163483849 User purchase article

### PR DESCRIPTION
**Pull request by Proception**
#### What does this PR do?
It implements the purchase functionality of users and artists being able to purchase an artwork and get the receipt for the payment

#### Description of Task to be completed?
- Modify the art model to add a price
- modify transaction model column 'amount' datatype to decimal
- add transaction validator
- add endpoint to return all transactions for a user
- add endpoint to return a transaction for an artwork
- add endpoint to purchase an artwork and make it unavailable to other users

#### How should this be manually tested?
- clone branch
- run `npm i`
- run `npm test`
- run `npm run start:dev`
- With an authorized user send the following requests below
- Send a POST request to the endpoint `localhost:9000/api/v1/transaction/purchase/{artId}` to purchase an art
- Send a GET request to the endpoint `localhost:9000/api/v1/transaction/receipt/{artId}` to get a single receipt for an art
- Send a GET request to the endpoint `localhost:9000/api/v1/transaction/` to get all transactions for a user both as a seller and buyer

#### Any background context you want to provide?
- An Item cannot be bought twice
- An artist/user cannot purchase his own art
#### What are the relevant pivotal tracker stories?
[#163483849](https://www.pivotaltracker.com/story/show/163483849)

#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

File: src/controllers/TransactionController.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+     * @desc GET /api/v1/transaction
+     * @param {object} req
+     * @param {object} res
+     * @memberof TransactionController
+     * @returns {object} res
+     */
+  static async getTransactions(req, res) {
+    const { id } = req.verifyUser;
+    try {
+      const allUserTxns = await transaction.getTransactions(id);
```
Comment (abejide001) [inline src/controllers/TransactionController.js]: please use Transactions here instead

File: src/db/models/art.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ module.exports = (sequelize, DataTypes) => {
     description: DataTypes.TEXT,
     categoryId: DataTypes.INTEGER,
     featuredImg: DataTypes.STRING,
+    price: DataTypes.DECIMAL(10, 2),
```
Comment (abejide001) [inline src/db/models/art.js:11]: can you please explain this line

**Review by abejide001 — CHANGES_REQUESTED** [review]

File: src/middlewares/TransactionValidator.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        return res.status(response.code).json(response);
+      }
+    }
+
+    /** Check Availability of product */
+    if (!req.validationErrors() && artObj) {
+      const { status } = artObj;
+      const isAvailable = status === false;
+      try {
+        req.check('verifyArt', 'Art is not available for sale')
```
Comment (wombolo) [inline src/middlewares/TransactionValidator.js:46]: Kindly explain what this line does

File: src/controllers/TransactionController.js (hunk 1 above)
Comment (Proception) [reply src/controllers/TransactionController.js]: done

File: src/db/models/art.js (hunk 2 above)
Comment (Proception) [reply src/db/models/art.js:11]: It means save the data in two decimal points. This is mostly used fields that hold currencies

**Review by wombolo — APPROVED** [review]
LGTM

File: src/middlewares/TransactionValidator.js (hunk 3 above)
Comment (Proception) [reply src/middlewares/TransactionValidator.js:46]: It checks the status of an art. If false, 'isAvailable' holds true. Then the custom validator needs 'false' as a response to throw an error so we negate it to false and vice versa when the status is true.

File: src/db/models/art.js (hunk 2 above)
Comment (abejide001) [reply src/db/models/art.js:11]: okay

**Review by abejide001 — APPROVED** [review]

**Review by julietezekwe — APPROVED** [review]

**Pull request merged**